File: backend/TaskManager.API/GraphQL/Mutations/Projects/ProjectRoleMutations.cs
using HotChocolate.Authorization;
using System.Security.Claims;
using TaskManager.API.Repositories.ProjectsRepository;

namespace TaskManager.API.GraphQL.Mutations.Projects
{
    [ExtendObjectType(OperationTypeNames.Mutation)]
    public class ProjectRoleMutations
    {
        [Authorize]
        public async Task<bool> UpdateProjectMemberRoleAsync(
            Guid projectId,
            Guid userId,
            string projectRole,
            Guid? roleLabelId,
            [Service] IProjectRepository projectRepository,
            ClaimsPrincipal claimsPrincipal)
        { 
            var currentUserIdStr = claimsPrincipal.FindFirst(ClaimTypes.NameIdentifier)?.Value?? claimsPrincipal.FindFirst("sub")?.Value;
            if (!Guid.TryParse(currentUserIdStr, out var currentUserId))
            {
                throw new GraphQLException("Не вдалось авторизувати користувача.");
            }

            if (currentUserId == userId)
            {
                throw new GraphQLException("Ви не можете змінювати власну роль у проєкті.");
            }

            if (projectRole is not ("manager" or "contributor"))
            {
                throw new GraphQLException("Невірна роль проєкту.");
            }

            var success = await projectRepository.UpdateProjectMembershipRoleAsync(
                projectId,
                userId,
                projectRole,
                roleLabelId,
                currentUserId
            );

            if (!success)
            {
                throw new GraphQLException("Не вдалося оновити роль користувача в проекті.");
            }

            return true;

        }
    }
}

File: frontend/task-frontend-manual/src/graphql/mutations/project/projectRoleMutations.ts
import { gql } from "@apollo/client";

export const UPDATE_PROJECT_MEMBER_ROLE = gql`
  mutation UpdateProjectMemberRole(
    $projectId: UUID!
    $userId: UUID!
    $projectRole: String!
    $roleLabelId: UUID
  ) {
    updateProjectMemberRole(
      projectId: $projectId
      userId: $userId
      projectRole: $projectRole
      roleLabelId: $roleLabelId
    )
  }
`;

export const CREATE_PROJECT_ROLE = gql`
    mutation CreateProjectRole($projectId: UUID!, $name: String!) {
        createProjectRole(projectId: $projectId, name: $name) {
            id
            name
        }
    }
`;